core/common/src/main/java/alluxio/resource/RoundRobinResourceQueue.java

```diff
@@ -0,0 +1,176 @@
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This class implements a ResourceQueue. A queue manages resources under a different
+ * model than a {@link Pool}: in a pool, resources are owned exclusively, and the workflow is:
+ * user acquires a resource from the pool, owns it and work on it exclusively for a time, and then
+ * return it back to the pool so that others can acquire it.
+ * <p></p>
+ * A resource queue manages <b>Shared Resources</b>, i.e., a list of resources shared among all
+ * users. When users acquire a resource from queue, they might get a new one, or they might get
+ * a shared one with other users. The idea is that the underlying resources can be safely shared
+ * so that acquire never needs to wait or fail.
+ * <p></p>
+ * {@link RoundRobinResourceQueue} implements the resource queue model by maintaining a queue of
+ * resources, each time allocating the first item in the queue and move it to the last, thereby
+ * simulating a round-robin mechanism.
+ *
+ * @param <T> The inner resource type
+ */
+@ThreadSafe
+public abstract class RoundRobinResourceQueue<T> implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(RoundRobinResourceQueue.class);
+
+  /** Max number of resources the queue can hold. */
+  private final int mMaxCapacity;
+
+  /** Lock protects internal mutation of resource queue. */
+  private final ReentrantLock mLock = new ReentrantLock();
+
+  /**
+   * Resources are wrapped in a {@link CountingReference} to keep track of its users.
+   * All the resources are managed in a deque, so that users acquire resource from the
+   * head of the deque, and newly allocated resource are moved to the back of the deque.
+   */
+  private final Deque<CountingReference<T>> mResourceQueue = new ArrayDeque<>();
+
+  /** GC executor that closes unused resources periodically. */
+  private final ScheduledExecutorService mExecutor;
+
+  /** GC task future. */
+  private final ScheduledFuture<?> mGcFuture;
+
+  /**
+   * Create a new resource queue.
+   * @param options construction options
+   */
+  public RoundRobinResourceQueue(DynamicResourcePool.Options options) {
+    mExecutor = Preconditions.checkNotNull(options.getGcExecutor());
+    mMaxCapacity = options.getMaxCapacity();
+    mGcFuture = mExecutor.scheduleAtFixedRate(
+        this::gcUnusedResources,
+        options.getInitialDelayMs(),
+        options.getGcIntervalMs(),
+        TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Acquire a {@link CloseableResource} from the queue.
+   * This method never fails or waits. If there is available slots in the queue,
```

Comment:
What if `createNewResource` fails?

Reply:
I haven't thought about how to deal with this situation. Currently if there's available slots in the queue, but the `createNewResource` fails, the method fails as a whole. It might make sense to fall back to allocating a shared one in that situation. But that complicates control flow, and in fact constructing a client never fails. Any suggestions? Can discuss it in more details.